dbms/src/Parsers/ParserCreateQuery.cpp

```diff
@@ -541,4 +544,185 @@ bool ParserCreateQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
+    ParserKeyValuePairsList expression_parser;
+
+    ASTPtr primary_key;
+    ASTPtr composite_key;
+    ASTPtr ast_source;
+    ASTPtr ast_lifetime;
+    ASTPtr ast_layout;
+    ASTPtr ast_range;
+
+    if (primary_key_keyword.ignore(pos) && !expression_list_p.parse(pos, primary_key, expected))
+        return false;
+
+    if (composite_key_keyword.ignore(pos) && !key_value_pairs_list_p.parse(pos, composite_key, expected))
+        return false;
+
+    /// Exactly one of two keys should be defined
+    if (!(bool(primary_key) ^ bool(composite_key)))
+        return false;
+
+    if (composite_key_keyword.ignore(pos))
+        expression_parser.parse(pos, composite_key, expected);
+
+    /// Exactly one of two keys should be defined
+    if (!(bool(primary_key) ^ bool(composite_key)))
+        return false;
+
+    while (true)
+    {
+        if (!ast_source && source_keyword.check_without_moving(pos, expected))
+        {
+            if (!key_value_pairs_p.parse(pos, ast_source, expected))
+                return false;
+            continue;
+        }
+
+        if (!ast_lifetime && lifetime_keyword.check_without_moving(pos, expected))
+        {
+            if (!key_value_pairs_p.parse(pos, ast_lifetime, expected))
+                return false;
+            continue;
+        }
+
+        if (!ast_layout && layout_keyword.check_without_moving(pos, expected))
+        {
+            if (!key_value_pairs_p.parse(pos, ast_layout, expected))
+                return false;
+            continue;
+        }
+
+        if (!ast_range && range_keyword.check_without_moving(pos, expected))
+        {
+            if (!key_value_pairs_p.parse(pos, ast_range, expected))
+                return false;
+            continue;
+        }
+
+        break;
+    }
+
+    auto query = std::make_shared<ASTDictionarySource>();
+    node = query;
+    if (primary_key)
+        query->set(query->primary_key, primary_key);
+
+    if (composite_key)
+        query->set(query->composite_key, composite_key);
+
+    if (ast_source)
+        query->set(query->source, ast_source);
+
+    if (ast_lifetime)
+        query->set(query->lifetime, ast_lifetime);
+
+    if (ast_layout)
+        query->set(query->layout, ast_layout);
+
+    if (ast_range)
+        query->set(query->range, ast_range);
+
+    return true;
+}
+
+
+const char * ParserCreateDictionaryQuery::getName() const
+{
+    return "CREATE DICTIONARY";
+}
+
+
+bool ParserCreateDictionaryQuery::parseImpl(IParser::Pos &pos, ASTPtr &node, Expected &expected)
+{
+    ParserKeyword s_create("CREATE");
+    ParserKeyword s_dictionary("DICTIONARY");
+    ParserKeyword s_if_not_exists("IF NOT EXISTS");
+    ParserKeyword s_or_replace("OR REPLACE");
+    ParserIdentifier name_p;
+    ParserToken s_left_paren(TokenType::OpeningRoundBracket);
+    ParserToken s_right_paren(TokenType::ClosingRoundBracket);
+    ParserToken s_dot(TokenType::Dot);
+    ParserColumnsOrIndicesDeclarationList columns_p(true);
```

Comment:
What about additional attributes on columns https://clickhouse.yandex/docs/en/query_language/dicts/external_dicts_dict_structure/#attributes?

Reply:
First argument of ParserColumnsOrIndicesDeclarationList constructor responsible for parsing that attributes. However, there are problems with interpretations of values. [Here](https://github.com/yandex/ClickHouse/pull/4178/files#diff-039e88a0be182d9489b59c36198c4e1bR510) I need a little smarter extraction of value.